dynatrace/stackstate_checks/dynatrace/dynatrace.py

```diff
@@ -0,0 +1,439 @@
+# (C) StackState 2021
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import time
+
+try:
+    from urllib.parse import urljoin
+except ImportError:
+    from urlparse import urljoin
+
+from datetime import datetime, timedelta
+
+from requests import Session, Timeout
+from schematics import Model
+from schematics.types import IntType, URLType, StringType, ListType, BooleanType, ModelType, DictType
+
+from stackstate_checks.base import AgentCheck, StackPackInstance
+from stackstate_checks.utils.identifiers import Identifiers
+
+VERIFY_HTTPS = True
+TIMEOUT_DEFAULT = 10
+EVENTS_BOOSTRAP_DAYS_DEFAULT = 5
+EVENTS_PROCESS_LIMIT_DEFAULT = 10000
+
+dynatrace_entities_cache = {}
+
+TOPOLOGY_API_ENDPOINTS = {
+    "process": "api/v1/entity/infrastructure/processes",
+    "host": "api/v1/entity/infrastructure/hosts",
+    "application": "api/v1/entity/applications",
+    "process-group": "api/v1/entity/infrastructure/process-groups",
+    "service": "api/v1/entity/services"
+}
+
+
+class DynatraceEvent(Model):
+    eventId = IntType()
+    startTime = IntType()
+    endTime = IntType()
+    entityId = StringType()
+    entityName = StringType()
+    severityLevel = StringType()
+    impactLevel = StringType()
+    eventType = StringType()
+    eventStatus = StringType()
+    tags = ListType(DictType(StringType))
+    id = StringType()
+    source = StringType()
+
+
+class State(Model):
+    last_processed_event_timestamp = IntType(required=True)
+
+
+class InstanceInfo(Model):
+    url = URLType(required=True)
+    token = StringType(required=True)
+    instance_tags = ListType(StringType, default=[])
+    events_boostrap_days = IntType(default=EVENTS_BOOSTRAP_DAYS_DEFAULT)
+    events_process_limit = IntType(default=EVENTS_PROCESS_LIMIT_DEFAULT)
+    verify = BooleanType(default=VERIFY_HTTPS)
+    cert = StringType()
+    keyfile = StringType()
+    timeout = IntType(default=TIMEOUT_DEFAULT)
+    domain = StringType(default='dynatrace')
+    environment = StringType(default='production')
+    state = ModelType(State)
+
+
+class DynatraceCheck(AgentCheck):
+    INSTANCE_TYPE = "dynatrace"
+    SERVICE_CHECK_NAME = "dynatrace"
+    INSTANCE_SCHEMA = InstanceInfo
+
+    def get_instance_key(self, instance_info):
+        return StackPackInstance(self.INSTANCE_TYPE, str(instance_info.url))
+
+    def check(self, instance_info):
+        try:
+            if not instance_info.state:
+                # Create state on the first run
+                empty_state_timestamp = self._generate_bootstrap_timestamp(instance_info.events_boostrap_days)
+                self.log.debug('Creating new empty state with timestamp: %s', empty_state_timestamp)
+                instance_info.state = State({'last_processed_event_timestamp': empty_state_timestamp})
+            self.start_snapshot()
+            self._process_topology(instance_info)
+            self._process_events(instance_info)
+            self.stop_snapshot()
```

Comment:
@hruhek Stop the snapshot before processing the events. I think this might solve the problem as what you are facing right now.

Reply:
@vishalxebia I didn't move `process_events` as I didn't obeserve any difference is STS regarding events or health checks in my testing. I'll move `process_events` after `stop_snapshot` if you think that should be done.